### PR TITLE
mount: add BindOptions.NonRecursive (API v1.40)

### PR DESCRIPTION
**Pull request by AkihiroSuda**
This allows non-recursive bind-mount, i.e. mount(2) with "bind" rather than "rbind".

Signed-off-by: Akihiro Suda <suda.akihiro@lab.ntt.co.jp>

**- What I did**

Support non-recursive bind-mount.
Related to https://github.com/moby/moby/issues/37838 .

**- How I did it**

* Added `NonRecursive` to `BindOptions`.
* Bumped up API version to v1.40
* Swarm-mode will be supported via a separate PR, because of mutual vendoring

**- How to verify it**

CLI: https://github.com/docker/cli/pull/1430
```console
$ mountpoint /run
/run is a mountpoint
$ docker run --rm --mount type=bind,src=/,target=/host,readonly busybox touch /host/run/compromise_success
$ docker run --rm --mount type=bind,src=/,target=/host,readonly,bind-nonrecursive busybox touch /host/run/compromise_fail
touch: /host/run/compromise_fail: Read-only file system
$ sudo ls /run/compromise_*
/run/compromise_success
```

**- Description for the changelog**

mount: add BindOptions.NonRecursive (API v1.40)

**- A picture of a cute animal (not mandatory but encouraged)**
https://upload.wikimedia.org/wikipedia/commons/2/28/Kaiserpinguine_mit_Jungen.jpg

penguins

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @justincormack @kolyshkin ptal

File: docs/api/version-history.md
Comment (thaJeztah) [inline docs/api/version-history.md]: Can you also mention;

- `GET /containers/<id>/json`
- `GET /containers/json`

And when swarmkit changes are done;

- `POST /services/create`
- `POST /services/{id}/update`
- `GET /services`
- `GET /services/{id}`

File: api/types/mount/mount.go
Comment (thaJeztah) [inline api/types/mount/mount.go]: Does `omitempty` actually work for a `bool` (not a pointer)? If not; perhaps we should make it a pointer so that it can be kept empty (and we can distinguish `false` from `not set`)

We should also make sure that this option is ignored on older API versions; i.e., if a request is sent to API version < 1.40, the `NoRecursive` option should be set to `nil` (or `false`); similar to https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/api/server/router/container/container_routes.go#L463-L466


Same for `createService`; https://github.com/moby/moby/blob/9f296d1e6fccd9f73bae345e4aad4f3f6e92fdeb/api/server/router/swarm/cluster_routes.go#L189-L195

And for `updateService`;

https://github.com/moby/moby/blob/9f296d1e6fccd9f73bae345e4aad4f3f6e92fdeb/api/server/router/swarm/cluster_routes.go#L232-L238

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Left some comments about the API changes; I'm not too familiar with the actual changes for mounting, so I'll leave that to @cpuguy83 and @kolyshkin

File: api/types/mount/mount.go
Comment (AkihiroSuda) [reply api/types/mount/mount.go]: https://play.golang.org/p/gEZoDu7LdG

> Does omitempty actually work for a bool (not a pointer)? 

Seems yes

**Comment by AkihiroSuda** [comment]
Updated PR. (Swarm-mode will be separate PR)

**Comment by olljanat** [comment]
@AkihiroSuda what is status of this one? My PR #37940 also contains some changes which will be available on API version 1.40 so it would be nice this one which prepares it would be merged first.

**Comment by olljanat** [comment]
@AkihiroSuda what is status of this one? My PR #37940 also contains some changes which will be available on API version 1.40 so it would be nice this one which prepares it would be merged first.

File: api/server/router/container/container_routes.go
Comment (cpuguy83) [inline api/server/router/container/container_routes.go]: Are we doing this sort of change on other fields?

File: api/swagger.yaml
Comment (cpuguy83) [inline api/swagger.yaml]: Maybe `NonRecursive`

Comment (AkihiroSuda) [reply api/swagger.yaml]: It was in my mind but  `NoRecursive` looks more consistent with volume `NoCopy`?

File: api/server/router/container/container_routes.go
Comment (AkihiroSuda) [reply api/server/router/container/container_routes.go]: https://github.com/moby/moby/pull/38003#discussion_r224488064

File: api/swagger.yaml
Comment (AkihiroSuda) [reply api/swagger.yaml]: how about NoRecursion

**Comment by AkihiroSuda** [comment]
@thaJeztah just to confirm, now this PR needs to be bumped up to v1.41, right?

**Comment by olljanat** [comment]
@AkihiroSuda I understand it on way that you just need just fix conflicts after #38089 is merged. Target API version for this one is still 1.4.0 which will be shipped part of 19.03.

**Comment by thaJeztah** [comment]
> @thaJeztah just to confirm, now this PR needs to be bumped up to v1.41, right?

@AkihiroSuda no, v1.40 is correct; we just forgot to bump the version before those other two PR's got merged. Docker 18.09 ships with API v1.39, so current master is / should be API v1.40

**Comment by thaJeztah** [comment]
Oh, yes, so what @olljanat says :joy:

**Comment by AkihiroSuda** [comment]
thanks, rebased

File: api/swagger.yaml
Comment (cpuguy83) [reply api/swagger.yaml]: I'm thinking "recursion" or "recursive" is not very descriptive here...
Wonder if we can come up with something different... some of the issue is we are mixing an action with a (linux mount) type.

Some ideas:

- `NoRetainSubmounts`
- `NoRetainMountStructure`
- `DisableSubmounts` (shorter, but not necessarily as clear as to what this means)

Comment (AkihiroSuda) [reply api/swagger.yaml]: @thajeztah @dmcgowan any thought?

Comment (thaJeztah) [reply api/swagger.yaml]: Naming is hard 😞 . Naming would probably have been easier if "non-recursive" was the default 😅 . I think "recursive" may actually fit the bill (as, e.g. `cp -r` , `rm -r` also use the "recursive" terminology). Slightly leaning towards `NonRecursive` over `NoRecursive` as well (but see where you're coming from with the `NoCopy` case).

I'll give it some thinking, maybe I can come up with alternatives (but it's hard!). `非再帰` 😉 😜 ?

Comment (AkihiroSuda) [reply api/swagger.yaml]: or `GeenRecursie` (if Google translation is correct) 😄

Comment (AkihiroSuda) [reply api/swagger.yaml]: Actually I'm fine with `NonRecursive`.

**Comment by AkihiroSuda** [comment]
renamed to `NonRecursive`

File: api/swagger.yaml
Comment (thaJeztah) [reply api/swagger.yaml]: > or `GeenRecursie` (if Google translation is correct) 😄

👍 🇳🇱 

> Actually I'm fine with `NonRecursive`

I'm fine with `NonRecursive` as well - @cpuguy83 fine with you as well?

File: daemon/volumes_unix.go
Comment (thaJeztah) [inline daemon/volumes_unix.go]: This is a bit hard to read; perhaps;

```suggestion
			var nonRecursive bool
			if m.Spec.Type == mounttypes.TypeBind && m.Spec.BindOptions != nil {
				nonRecursive = m.Spec.BindOptions.NonRecursive
			}
```

Or perhaps even better; move it lower, then the intermediate variable is not even needed;

```suggestion
			mnt := container.Mount{
				Source:      path,
				Destination: m.Destination,
				Writable:    m.RW,
				Propagation: string(m.Propagation),
				Source:       path,
				Destination:  m.Destination,
				Writable:     m.RW,
				Propagation:  string(m.Propagation),
			}
			if m.Spec.Type == mounttypes.TypeBind && m.Spec.BindOptions != nil {
				mnt.NonRecursive = m.Spec.BindOptions.NonRecursive
			}
```

Comment (thaJeztah) [reply daemon/volumes_unix.go]: Looking at the above; `NonRecursive = true` is only supported for `type=bind`, we should add that `ValidateMountConfig()`;

https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/volume/mounts/linux_parser.go#L50-L110

and LCOW / Windows;

https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/volume/mounts/lcow_parser.go#L10-L26

https://github.com/moby/moby/blob/8e610b2b55bfd1bfa9436ab110d311f5e8a74dcb/volume/mounts/windows_parser.go#L216-L299

File: daemon/volumes_unix.go
Comment (thaJeztah) [inline daemon/volumes_unix.go]: Would it be slightly clearer to use a variable for both "bind mode" and "write mode"? something like;

```suggestion
		bindMode := "rbind"
		if m.NonRecursive {
			bindMode = "bind"
		}

		writeMode := "ro"
		if m.Writable {
			writeMode = "rw"
		}

		opts := strings.Join([]string{bindMode, writeMode}, ",")
```

File: api/swagger.yaml
Comment (thaJeztah) [inline api/swagger.yaml]: We may want to add a bit more information about what this does (also, that it only applies to / can be used for `Type=bind`)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks @AkihiroSuda - I left some comments.

Erm... (not gonna make myself popular with you 😅) I was just recalling a discussion we had on another PR about options for `tmpfs`. (this PR: https://github.com/moby/moby/pull/36720)

Thinking now if we should have a dedicated boolean for this in the API, or take the same approach as was used in that PR.


@AkihiroSuda @cpuguy83 @kolyshkin wdyt?

**Comment by AkihiroSuda** [comment]
> Thinking now if we should have a dedicated boolean for this in the API, or take the same approach as was used in that PR.

I feel this PR is really specific to `BindOptions`

File: api/swagger.yaml
Comment (cpuguy83) [reply api/swagger.yaml]: This is already in `BindOptions`, I think that's good.

**Comment by AkihiroSuda** [comment]
updated

File: integration/container/mounts_linux_test.go
Comment (cpuguy83) [inline integration/container/mounts_linux_test.go]: Could probably get rid of the remote daemon check if we do all the mounting in a container with shared propagation.

**Review by thaJeztah — APPROVED** [review]
LGTM after @cpuguy83's comment https://github.com/moby/moby/pull/38003/files#r231328643 is addressed (but I'd be ok with keeping that for a follow-up)

**Comment by cpuguy83** [comment]
Yeah LGTM as is.
On mobile so I can't merge with failing ci.

File: integration/container/mounts_linux_test.go
Comment (AkihiroSuda) [reply integration/container/mounts_linux_test.go]: Could you be more specific? Do you mean DIND?

**Comment by AkihiroSuda** [comment]
Swarmkit PR: https://github.com/docker/swarmkit/pull/2780

**Comment by AkihiroSuda** [comment]
PR for Swarm-mode: https://github.com/moby/moby/pull/38788

**Pull request merged**
